File: cs50 assignments/credit.c
// credit.c checks if an inputted credit number is a valid credit card type. For specifications, view: https://cs50.harvard.edu/x/2022/psets/1/credit/


#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card = get_long("Number: ");
    int digits = 0;
    int prefix, remainder;
    int even = 0;
    int luhns = 0;

    // count digits in number
    do
    {
        remainder = card % 10;
        card = card / 10;
        digits++;

        // find card prefix
        if (card < 100 && card >= 10)
        {
            prefix = card;
        }

        // calculate Luhn's Algorithm p1
        if (digits % 2 == 0)
        {
            remainder = remainder * 2;
            remainder = remainder % 10 + remainder / 10;
            even = even + remainder;
        }

        // calculate Luhn's Algorithm p2
        else if (digits % 2 > 0)
        {
            luhns = luhns + remainder;
        }
    }
    while (card > 1);

    // find Luhn's #
    luhns = luhns + even;

    // What card is it
    if (luhns % 10 == 0)
    {
        // AMEX identifier
        if (digits == 15 && (prefix == 34 || prefix == 37))
        {
            //if (digits == 15)
            //{
                printf("AMEX\n");
            //}
            //else
            //{
                //printf("INVALID\n");
            //}
        }

        // MASTERCARD identifier
        else if (digits == 16 && (prefix >= 51 && prefix <= 55))
        {
            //if (digits == 16)
            //{
                printf("MASTERCARD\n");
            //}
            //else
            //{
                //printf("INVALID\n");
            //}
        }

        // VISA identifier
        else if ((digits == 13 || digits == 16) && (prefix >= 40 && prefix <= 49))
        {
            //if (digits == 13 || digits == 16)
            //{
                printf("VISA\n");
            //}
            //else
            //{
                //printf("INVALID\n");
            //}
        }

        // solves luhn's but not a card
        else
        {
            printf("INVALID\n");
        }
    }

    // not a card #
    else
    {
        printf("INVALID\n");
    }
}

File: cs50 assignments/credit.py
#14th Assignment. credit.py is a reimplementation of credit.c in python.

# TODO
from cs50 import get_string

# ask user for a credit card number
card = get_string("Number: ")
length = len(card)
luhns = 0
# calculate luhn's number starting from the second to last digit
for i in range(length):
    if i % 2 == 0:
        luhns = luhns + int(card[length - 1 - i])
    else:
        if int(card[length - 1 - i]) * 2 > 9:
            luhns = luhns + 1 + ((int(card[length - 1 - i]) % 10) * 2)
        else:
            luhns = luhns + (int(card[length - 1 - i]) * 2)
# if luhn's number does not end in 0, the card is invalid
if luhns % 10 == 0:
    # check which card brand or if invalid
    if length == 15 and int(card[0]) == 3 and (int(card[1]) == 4 or int(card[1]) == 7):
        print("AMEX")
    elif length == 16 and int(card[0]) == 5 and (int(card[1]) > 0 and int(card[1]) < 6):
        print("MASTERCARD")
    elif int(card[0]) == 4 and (length == 13 or length == 16):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
